src/transport/webrtc/opening.rs

```diff
@@ -0,0 +1,473 @@
+// Copyright 2023-2024 litep2p developers
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+//! WebRTC handshaking code for an opening connection.
+
+use crate::{
+    config::Role,
+    crypto::{ed25519::Keypair, noise::NoiseContext},
+    transport::{webrtc::util::WebRtcMessage, Endpoint},
+    types::ConnectionId,
+    Error, PeerId,
+};
+
+use multiaddr::{multihash::Multihash, Multiaddr, Protocol};
+use str0m::{
+    change::Fingerprint,
+    channel::ChannelId,
+    net::{DatagramRecv, DatagramSend, Protocol as Str0mProtocol, Receive},
+    Event, IceConnectionState, Input, Output, Rtc,
+};
+
+use std::{net::SocketAddr, time::Instant};
+
+/// Logging target for the file.
+const LOG_TARGET: &str = "litep2p::webrtc::connection";
+
+/// Create Noise prologue.
+fn noise_prologue(local_fingerprint: Vec<u8>, remote_fingerprint: Vec<u8>) -> Vec<u8> {
+    const PREFIX: &[u8] = b"libp2p-webrtc-noise:";
+    let mut prologue =
+        Vec::with_capacity(PREFIX.len() + local_fingerprint.len() + remote_fingerprint.len());
+    prologue.extend_from_slice(PREFIX);
+    prologue.extend_from_slice(&remote_fingerprint);
+    prologue.extend_from_slice(&local_fingerprint);
```

Comment:
IIUC, this is <prefix><fingerprint A><fingerprint B>, depending on whatever this is the handshake responder or the initator.
(extracted from: https://github.com/libp2p/specs/blob/master/webrtc/webrtc-direct.md#connection-security)

Do we need to handle the case where we are the responder? Ie <prefix><local><remote> instead of <prefix><remote><local>

Reply:
My guess is it's based on the role of smoldot in the WebRTC noise handshake. It looks like we don't support litep2p to litep2p WebRTC connections.

Reply:
This PR only implemented WebRTC Direct (browser <-> server), not server <-> server.